gdal/port/cpl_vsil_hdfs.cpp

```diff
@@ -0,0 +1,320 @@
+
+//! @cond Doxygen_Suppress
+
+#include <string>
+
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <cstring>
+
+#include "cpl_port.h"
+#include "cpl_vsi.h"
+
+#include "cpl_conv.h"
+#include "cpl_error.h"
+#include "cpl_vsi_virtual.h"
+
+#include "hdfs.h"
+
+CPL_CVSID("$Id$")
+
+const char * VSIHDFS = "/vsihdfs/";
+
+
+/************************************************************************/
+/* ==================================================================== */
+/*                        VSIHdfsHandle                               */
+/* ==================================================================== */
+/************************************************************************/
+
+class VSIHdfsHandle final : public VSIVirtualHandle
+{
+  private:
+    CPL_DISALLOW_COPY_ASSIGN(VSIHdfsHandle)
+
+    hdfsFile poFile = nullptr;
+    hdfsFS poFilesystem = nullptr;
+    std::string oFilename;
+    bool bReadOnly;
+
+  public:
+     VSIHdfsHandle(hdfsFile poFile,
+                   hdfsFS poFilesystem,
+                   const char * pszFilename,
+                   bool bReadOnly);
+    ~VSIHdfsHandle() override;
+
+    int Seek(vsi_l_offset nOffset, int nWhence) override;
+    vsi_l_offset Tell() override;
+    size_t Read(void *pBuffer, size_t nSize, size_t nMemb) override;
+    size_t Write(const void *pBuffer, size_t nSize, size_t nMemb) override;
+    vsi_l_offset Length()
+    {
+      hdfsFileInfo * poInfo = hdfsGetPathInfo(this->poFilesystem, this->oFilename.c_str());
+      if (poInfo != NULL) {
+        tOffset nSize = poInfo->mSize;
+        hdfsFreeFileInfo(poInfo, 1);
+        return static_cast<vsi_l_offset>(nSize);
+      }
+      return -1;
+    }
+    int Eof() override;
+    int Flush() override;
+    int Close() override;
+};
+
+VSIHdfsHandle::VSIHdfsHandle(hdfsFile _poFile,
+                             hdfsFS _poFilesystem,
+                             const char * pszFilename,
+                             bool _bReadOnly)
+  : poFile(_poFile), poFilesystem(_poFilesystem), oFilename(pszFilename), bReadOnly(_bReadOnly)
+{}
+
+VSIHdfsHandle::~VSIHdfsHandle()
+{
+  this->Close();
+}
+
+int VSIHdfsHandle::Seek(vsi_l_offset nOffset, int nWhence)
+{
+  switch(nWhence) {
+  case SEEK_SET:
+    return hdfsSeek(this->poFilesystem, this->poFile, nOffset);
+  case SEEK_CUR:
+    return hdfsSeek(this->poFilesystem,
+                    this->poFile,
+                    nOffset + this->Tell());
+  case SEEK_END:
+    return hdfsSeek(this->poFilesystem,
+                    this->poFile,
+                    static_cast<tOffset>(this->Length()) - nOffset);
+  default:
+    return -1;
+  }
+}
+
+vsi_l_offset VSIHdfsHandle::Tell()
+{
+  return hdfsTell(this->poFilesystem, this->poFile);
+}
+
+size_t VSIHdfsHandle::Read(void *pBuffer, size_t nSize, size_t nMemb)
+{
+  return hdfsRead(this->poFilesystem, this->poFile, pBuffer, nSize * nMemb);
```

Comment:
the method should return the number of members of size nSize read, not the number of bytes. So basically divide by nSize probably

Reply:
Done.